main.cpp

```diff
@@ -14,9 +14,139 @@
+			{"help", no_argument, 0, 'h'},
+			{"input-size", required_argument, 0, 's'},
+			{"input", required_argument, 0, 'i'},
+			{"output", required_argument, 0, 'o'},
+			{0, 0, 0, 0}};
+
+		int option_index = 0;
+		int c = getopt_long(argc, argv, "hs:i:o:", long_options, &option_index);
+		if (c == -1) {
+			break;
+		}
+		switch (c) {
+		case 'h':
+			PrintHelp();
+			return 0;
+
+		case 's': {
+			auto res = mender::common::StringToLongLong(optarg);
+			if (res) {
+				volumeSize = res.value();
+			} else {
+				std::cerr << res.error().message << std::endl;
+				;
+				PrintHelp();
+				exit(EXIT_FAILURE);
+			}
+			break;
+		}
+		case 'i':
+			inputPath = optarg;
+			break;
+
+		case 'o':
+			outputPath = optarg;
+			break;
+
+		case '?':
+			break;
+
+		default:
+			PrintHelp();
+			exit(EXIT_FAILURE);
+		}
+	}
+
+	if (inputPath.empty() || outputPath.empty()) {
+		std::cerr << "Wrong input parameters!" << std::endl;
+		PrintHelp();
+		exit(EXIT_FAILURE);
+	}
+
+	mender::io::File srcFile;
+	mender::io::File dstFile;
+
+	std::shared_ptr<mender::io::FileReader> reader;
+	if (inputPath == "-") {
+		srcFile = mender::io::GetInputStream();
+		reader = std::make_shared<mender::io::InputStreamReader>();
+	} else {
+		auto src = mender::io::Open(inputPath);
+		if (!src) {
+			std::cerr << "Failed to open source: " << inputPath << " (" << src.error().message
+					  << ")";
+			exit(EXIT_FAILURE);
+		}
+		srcFile = src.value();
+		reader = std::make_shared<mender::io::FileReader>(srcFile);
+	}
+
+	bool isUBI = false;
+	auto isUBIRes = mender::io::IsUBIDevice(outputPath);
+	if (isUBIRes.has_value()) {
+		isUBI = isUBIRes.value();
+	}
+
+	auto dst = mender::io::Open(
+		outputPath,
+		!isUBI, // read: only for non-UBI volumes
+		true);  // write
+
+	if (!dst) {
+		std::cerr << "Failed to open destination: " << outputPath << " (" << dst.error().message
+				  << ")";
+		exit(EXIT_FAILURE);
+	}
+	dstFile = dst.value();
+
+	if (isUBI) {
+		auto res = mender::io::SetUbiUpdateVolume(dstFile, volumeSize);
+		if (res != mender::common::error::NoError) {
+			std::cerr << res.message;
+			exit(EXIT_FAILURE);
+		}
+	}
+
+	mender::io::LimitedFlushingWriter flushWriter(dstFile, volumeSize, blockSize);
+	mender::io::FileWriter writer(dstFile);
+	mender::io::FileReadWriterSeeker readwriter(isUBI ? writer : flushWriter);
+	mender::OptimizedWriter optWriter(*reader, readwriter, blockSize, volumeSize);
+	optWriter.Copy(!isUBI);
+
+	auto statistics = optWriter.GetStatistics();
+
+	std::cout << "================ STATISTICS ================" << std::endl;
+	std::cout << "Blocks written: " << statistics.blocksWritten_ << std::endl;
+	std::cout << "Blocks omitted: " << statistics.blocksOmitted_ << std::endl;
+	std::cout << "Bytes  written: " << statistics.bytesWritten_ << std::endl;
+	std::cout << "============================================" << std::endl;
+
+	if (srcFile != mender::io::GetInputStream()) {
+		mender::io::Close(srcFile);
+	}
+	mender::io::Close(dstFile);
```

Comment:
Do you need `Close` here? Both `srcFile` and `dstFile` should be owned by the readers/writers at this point. Which was also the point, to not have to do this.

Reply:
I don't need, I forgot to remove these lines.